test/e2e/e2e_test.go

```diff
@@ -589,6 +589,96 @@ func pollForResource(t *testing.T, resourceID string, timeout time.Duration, acc
 	return obj, err
 }
 
+func checkClientPodRcvdUpdatedServerCert(t *testing.T, client *kubernetes.Clientset, service *v1.Service, testNS, updatedServerCert string) {
+	timeout := 5 * time.Minute
+	metricsURL := fmt.Sprintf("%s.%s.svc:%d", service.Name, service.Namespace, service.Spec.Ports[0].Port)
+	err := wait.PollImmediate(10*time.Second, timeout, func() (bool, error) {
+		podName := "client-pod-" + randSeq(5)
+		_, err := client.CoreV1().Pods(testNS).Create(&v1.Pod{
+			TypeMeta: metav1.TypeMeta{},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      podName,
+				Namespace: testNS,
+			},
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{
+					{
+						Name:    "cert-checker",
+						Image:   "nicolaka/netshoot:latest",
+						Command: []string{"/bin/bash"},
+						Args:    []string{"-c", fmt.Sprintf("openssl s_client -showcerts -connect %s < /dev/null 2>/dev/null | openssl x509", metricsURL)},
+					},
+				},
+				RestartPolicy: v1.RestartPolicyOnFailure,
+			},
+		})
+		defer deletePod(client, podName, testNS)
+		if err != nil {
+			t.Logf("creating client pod failed: %v", err)
+			return false, nil
+		}
+
+		err = waitForPodPhase(t, client, podName, testNS, v1.PodSucceeded)
+		if err != nil {
+			t.Logf("wait on pod to complete failed: %v", err)
+			return false, nil
+		}
+
+		serverCertClientReceived, err := getPodLogs(t, client, podName, testNS)
+		if err != nil {
+			t.Logf("fetching pod logs failed: %v", err)
+			return false, nil
+		}
+		return strings.Contains(updatedServerCert, serverCertClientReceived), nil
+	})
+	if err != nil {
+		t.Fatalf("failed to verify updated certs within timeout(%v)", timeout)
+	}
+
+}
+
+func waitForPodPhase(t *testing.T, client *kubernetes.Clientset, name, namespace string, phase v1.PodPhase) error {
+	return wait.PollImmediate(10*time.Second, time.Minute, func() (bool, error) {
```

Comment:
Nesting wait loops is something of an anti-pattern. Is it necessary in this case?

Reply:
this is something that out of our previous discussion of waiting till the pod is ready before the test fetches the logs. it simplifies the read for pod logs.